### ISSUE TITLE
Calling drain on async subscription dropping messages

### ISSUE THREAD

**Issue by JankyGaming** [nats-io/nats.go#898]
## Defect

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included nats.go version
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats.go` and the `nats-server` if one was involved:
nats-server: Current latest (2.7.1)?
nats.go: github.com/nats-io/nats.go v1.13.1-0.20220121202836-972a071d373d

#### OS/Container environment:
Mac OS for client
Current latest (2.7.1) standard docker image (alpine?) for server

#### Steps or code to reproduce the issue:
[Gist with tests](https://gist.github.com/JankyGaming/a763281e26d1b4434f47c14a293d10e1)

#### Expected result:
Expect all tests to pass, reflecting a 100% processing rate assuming that I am properly draining individual subscriptions.

#### Actual result:
near 100% test failures with various scenarios. (Calling Drain inside callback handler for sub seems to provide better, but still fallable results)

Hopefully, I am just ignorant to something and I am just improperly draining a subscription.

**Comment by JankyGaming** [comment]
This issue also occurs when calling drain on the connection.

**Comment by kozlovic** [comment]
There is a problem with the way you create the JS consumer. Since you pass almost all arguments but with 0 values, you are not creating the JS consumer the way you think, I guess. Here how the consumer should be created IMO:
```
_, err = jsc.AddConsumer(streamName, &nats.ConsumerConfig{
		Durable:        serviceName,
		Description:    "A consumer created by the seller service to consume new order events",
		DeliverSubject: strings.Join([]string{serviceName, streamName}, "."),
		DeliverGroup:   serviceName,
		FilterSubject:  "something.>",
		DeliverPolicy:  nats.DeliverAllPolicy,
		AckPolicy:      nats.AckExplicitPolicy,
		AckWait:        time.Second * 5,
	})
```
Notice the AckPolicy, you had it at 0, which means that it would be AckNone, which means that as soon as the server sends a message to the client, the server automatically consider that message acknowledged.
I also set the DeliverPolicy to make sure that you receive all messages. This one is equivalent of (incorrectly) setting to 0, but that makes it more clear what the intent it.

Then, your code needs to take into consideration that a message may have to be redelivered, so there could be a loop iteration where no new message is received, but it should not fail then. Since the AckWait is 5 seconds, you need to give a chance to the server to redeliver that message. What you could do is that if no new message is received in an iteration (but you didn't get all expected messages) is to increase the duration for the next loop, for a certain amount of tries.

I do see something in the client library that should be fixed and will post a PR for that, but with the information given above, even with library without a "fix", you should end-up getting all messages. Let me know if that is not the case.

**Comment by kozlovic** [comment]
Actually, I take back the comment about issue in the library. I was concerned about this cod:
```
if drainMode {
    go nc.checkDrained(sub)
}

// We will send these for all subs when we reconnect
// so that we can suppress here.
if !nc.isReconnecting() {
    nc.bw.appendString(fmt.Sprintf(unsubProto, s.sid, maxStr))
    nc.kickFlusher()
}
```
namely, the fact that we were starting the check for drained *before* we had a chance to enqueue the UNSUB protocol. But since the code above is done under the connection's lock and that checkDrained() first tries to flush the connection, which requires the lock, we are guaranteed to have the UNSUB protocol enqueued before checkDrained() can flush and check for drain status. So there is no changes needed in the client library.

**Comment by JankyGaming** [comment]
Ok great. It does seem that setting AckPolicy to AckExplicitPolicy is provided the behavior I'd expect.

Maybe you can explain how I should be interpreting the information in the documentation here @kozlovic . [https://docs.nats.io/using-nats/developer/receiving/drain](https://docs.nats.io/using-nats/developer/receiving/drain). The way it is explained there, I interpreted it to mean it would finish passing all already sent/in flight messages to the client through the callback/handler or the channel depending on sub type. To me I would assume no extra configuration would be needed per this documentation. I assume the reason the suggested configuration works is because calling Drain() does not actually "Process" all the messages (ie doesn't send them all to the cb), which is where an automatic ack occurs.

**Comment by kozlovic** [comment]
When a NATS (not JS) subscription is Drain()'ed, what this simply means is that the library will send an UNSUB protocol to the server, then send a PING and wait for corresponding PONG to ensure that the server has processed the UNSUB, therefore once all messages from the subscription have been drained, the library knows that no more are coming and the subscription can be closed. Of course, published messages may be incoming at the same time that the server receives the UNSUB. In core-NATS, your application has no control of that and of course you will miss messages that are published at that time and after.

With JetStream, while the server is delivering message to the NATS subscription attached to the JetStream consumer, the processing of the UNSUB for the low-level subscription, when it occurs, notifies the go routine that deliver messages to consumers that the interest is gone. This can't be "atomic" and therefore there is a window where the server does not realize yet that the subscription is gone and so will "deliver", that is, send and mark the message as pending, even if at the end the message does not reach the client. With an AckNone, that message is considered acknowledged in that process and your application may detect it as "missing". With other delivery mode, since this message will be marked as pending, and the server will not receive an ACK, it will be redelivered.

In summary, the Drain() is a core-NATS functionality, not a JetStream one.

**Comment by JankyGaming** [comment]
So can you explain to me what the documentation means when it says "it will process any inflight and cached/pending messages before closing.", particularly the bit about processing "inflight" messages? I would have assumed this meant that when interest is cleared, any messages still being sent to the client are what is called "Inflight" here.

**Comment by kozlovic** [comment]
> it will process any inflight and cached/pending messages before closing.

Suppose a connection receives for a given subscriptions messages: M1, M2, M3, M4, M5.
M1 is currently presented to the message callback. What is considered cached/pending are messages M2 to M5: they have been received by the connection and added to the internal subscription's list, waiting to be "dispatched" to the user callback.

When the user calls Drain(), what the library does is send an UNSUB protocol to the server. Before the server receives this UNSUB, it is possible that it was sending messages M6, M7 and M8: they may have been "added" to the socket buffer or are in transit on the network...

The server then processes the UNSUB and after that point will stop sending new messages.

As part of checking the drain status, the library issues a Flush(), which is a PING for which it waits for the corresponding PONG. When receiving the PONG, it ensures that the server has processed the UNSUB and that in the above example, that the library received messages M6, M7 and M8 have also been processed in the client connection, that is, added to the subscription's list, since they would be *before* the PONG response from the server.

The Drain() operation then waits for the number of pending messages to fall to 0 (or the connection/subscription being forcibly closed by the user) before closing the subscription, ensuring that pending messages M2 to M8 have been presented to the user's callback.

Let me know if this is still not clear.

**Comment by kozlovic** [comment]
@JankyGaming Because I explained that the Drain is core-NATS, not JetStream. See my previous comment regarding JetStream:
--
With JetStream, while the server is delivering message to the NATS subscription attached to the JetStream consumer, the processing of the UNSUB for the low-level subscription, when it occurs, notifies the go routine that deliver messages to consumers that the interest is gone. This can't be "atomic" and therefore there is a window where the server does not realize yet that the subscription is gone and so will "deliver", that is, send and mark the message as pending, even if at the end the message does not reach the client. With an AckNone, that message is considered acknowledged in that process and your application may detect it as "missing". With other delivery mode, since this message will be marked as pending, and the server will not receive an ACK, it will be redelivered.
--

In summary, there is a window of time between the processing on the UNSUB (on the core NATS subscription that the client library created on the deliver subject) and the time the server stops sending JS messages.
In the example above, it could be that the server process UNSUB after messages M8 but then would still pick messages M9 and M10 to be delivered, before the JS consumer loop has detected that it should stop sending messages. This is why I wrote that this detection between UNSUB and sending JS messages is not "atomic" (and likely can't be done with chosen architecture). 

In the server here is the place where messages are "sent" to the subscription:
https://github.com/nats-io/nats-server/blob/d71bb63e1bcfd213f5c556b697bd63c746c16590/server/consumer.go#L2750
and here is the place where it gets notified that the interest is gone: https://github.com/nats-io/nats-server/blob/d71bb63e1bcfd213f5c556b697bd63c746c16590/server/consumer.go#L2787
That is not "atomic" and can't really be made that way, with given architecture.

So the Drain() is really for your app to be able to process any message that has already arrived at the connection level and not "waste" those messages by ignoring them. But with JetStream and acknowledgement mode, the ones that have been "falling through the cracks" are not lost since they won't be acknowledged and they will be resent.

**Comment by JankyGaming** [comment]
Ok, for what it is worth I do think I understand what is going on here now. I would suggest updating the documentation to note that AckNonePolicy (The default of a subscription created in go) still could result in lost messages. The documentation implies NO lost messages with Drain() (vs Unsubscribe()) but does not present any description of the underlying subscriptions config. 

Thanks again!
